### PR TITLE
utils: add the systemd-dbus-api cpuset bitmask output

### PR DESCRIPTION
**Pull request by npf**
Add Systemd's D-Bus API cpuset output to `hwloc-calc`.

Example:
```
hwloc-calc --cpuset-input-format list 3-6,10 --cpuset-output-format systemd-dbus-api
AllowedCPUs ay 0x02 0x78 0x04
```

Example of use with Systemd's D-Bus API:
```
# Create the systems slice
busctl call org.freedesktop.systemd1 \
  /org/freedesktop/systemd1 \
  org.freedesktop.systemd1.Manager \
  StartUnit ss my_slice.slice fail

# Configure the CPU set of the slice thanks to hwloc-calc for the syntax translation 
busctl call org.freedesktop.systemd1 \
  /org/freedesktop/systemd1 \
  org.freedesktop.systemd1.Manager \
  SetUnitProperties 'sba(sv)' \
  my_slice.slice 1 1
  $(hwloc-calc pu:0 pu:31 pu:32 pu:63 pu:64 pu:77
--cpuset-output-format systems)

# Add the current process to the slice
busctl call org.freedesktop.systemd1 \
  /org/freedesktop/systemd1 \
  org.freedesktop.systemd1.Manager \
  StartTransientUnit 'ssa(sv)a(sa(sv))' \
  my_scope.scope fail 3 \
  Delegate b 1
  PIDs au 1 $$ \
  Slice s my_slice.slice \
  0
```

### CODE REVIEW

File: utils/hwloc/misc.h [hunk 1]
```diff
@@ -946,20 +947,69 @@ hwloc_utils_parse_cpuset_format(const char *string)
...
+ *   "AllowedCPUs ay 0x0a 0x01 0x00 0x00 x80 0x01 0x00 0x00 0x80 0x01 0x20
+ * */
+#define HWLOC_SYSTEMD_ALLOWEDCPUS_PREFIX_FORMAT "AllowedCPUs ay 0x%02x"
+#define HWLOC_SYSTEMD_ALLOWEDCPUS_PREFIX_SIZE 19
+#define HWLOC_SYSTEMD_ALLOWEDCPUS_BYTES_FORMAT " 0x%02x"
+#define HWLOC_SYSTEMD_ALLOWEDCPUS_BYTES_SIZE 5
+static __hwloc_inline int hwloc_utils_systemd_asprintf(char ** strp, const struct hwloc_bitmap_s * __hwloc_restrict set)
+{
+  int last = hwloc_bitmap_last(set);
+  if (last == -1 )
```
Comment (bgoglin) [inline utils/hwloc/misc.h:973]: abort() is too strong in this case. what should we do if the cpuset is empty? return an error and document that empty cpuset aren't supported?

File: utils/hwloc/misc.h [hunk 2]
```diff
@@ -946,20 +947,69 @@ hwloc_utils_parse_cpuset_format(const char *string)
...
+    abort();
+  int bytes = last / 8 + 1; /* how many bytes to represent the cpuset bit mask */
+  int buflen = HWLOC_SYSTEMD_ALLOWEDCPUS_PREFIX_SIZE + HWLOC_SYSTEMD_ALLOWEDCPUS_BYTES_SIZE * bytes + 1;
+  int ret = 0;
+#ifdef HWLOC_SYSTEMD_ALLOWEDCPUS_DEBUG
```
Comment (bgoglin) [inline utils/hwloc/misc.h:978]: is there any way to activate this debug? should we keep it? or just make it a comment? some compiler will complain about fprintf is the middle of variable declarations

File: utils/hwloc/hwloc-calc.1in [hunk 3]
```diff
@@ -534,6 +536,23 @@ OS devices may also be filtered by subtype. In this example, there are
     $ utils/hwloc/hwloc-calc -I coproc node:1
     7,8
 
+Example of use of the systemd-dbus-api cpuset output format:
```
Comment (bgoglin) [inline utils/hwloc/hwloc-calc.1in:539]: Start a new man section, the EXAMPLES one is already too long. Something like EXAMPLES - SYSTEMD DBUS API. I split the rest of examples later.

**Review by bgoglin — COMMENTED** [review]
Also add your copyright at the top of modified files, and a Signed-off-by in the commit log.

File: utils/hwloc/misc.h [hunk 4]
```diff
@@ -946,20 +947,69 @@ hwloc_utils_parse_cpuset_format(const char *string)
...
+      fprintf(stderr, "byte = %d => ith_ulong[%d] = 0x%016lx\n", byte, byte / HWLOC_SIZEOF_UNSIGNED_LONG, ith);
+#endif
+    }
+    ret += snprintf(buf + ret, HWLOC_SYSTEMD_ALLOWEDCPUS_BYTES_SIZE + 1, HWLOC_SYSTEMD_ALLOWEDCPUS_BYTES_FORMAT, (unsigned char) (ith & 0xFF));
+    ith >>= 8;
+  }
+#ifdef HWLOC_SYSTEMD_ALLOWEDCPUS_DEBUG
+  fprintf(stderr, "Left ith bytes = 0x%02lx\n", ith);
+#endif
+  assert(ith == 0); /* whenever some bytes of the ulong ith may not be consummed, they must be 0 */
```
Comment (bgoglin) [inline utils/hwloc/misc.h:1000]: "consumed"

File: utils/hwloc/misc.h (hunk 1 above)
Comment (npf) [reply utils/hwloc/misc.h:973]: For the systemd-dbus-apo case, it makes no sense not to pass a cpuset, so we could report an error: `Empty and infinite CPU sets are not supported with the systemd-dbus-api cpuset output format`. 

Can you confirm that `-1` can be any of both: empty or infinite?

Comment (bgoglin) [reply utils/hwloc/misc.h:973]: Yes, both empty and infinite.

File: utils/hwloc/misc.h (hunk 2 above)
Comment (bgoglin) [reply utils/hwloc/misc.h:978]: Nobody's ever going to enable them, I'd remove them and just make sure we have tests for different cases in make check.

File: utils/hwloc/hwloc-calc.1in (hunk 3 above)
Comment (npf) [reply utils/hwloc/hwloc-calc.1in]: I pushed some changes.

File: utils/hwloc/misc.h (hunk 1 above)
Comment (npf) [reply utils/hwloc/misc.h]: Fixed

File: utils/hwloc/misc.h (hunk 2 above)
Comment (npf) [reply utils/hwloc/misc.h]: done.

**Comment by npf** [comment]
> Also add your copyright at the top of modified files, and a Signed-off-by in the commit log.
Done

**Comment by bgoglin** [comment]
Thanks, I'll backport this to 2.x and 2.11 since it pretty much doesn't touch the existing code.

**Pull request merged**